pages/catalog/CatalogHandler.inc.php

```diff
@@ -51,21 +51,91 @@ function authorize($request, &$args, $roleAssignments) {
 	 * @param $request PKPRequest
 	 */
 	function index($args, $request) {
+		$this->page($args, $request, true);
+	}
+
+	/**
+	 * Show one page from the catalog
+	 * @param $args array
+	 * @param $request Request
+	 * @param $isFirstPage boolean
+	 */
+	public function page($args, $request, $isFirstPage) {
+		$page = null;
+		if ($isFirstPage) {
+			$page = 1;
+		} elseif ($args[0]) {
+			$page = (int) $args[0];
+		}
+
+		if (!$isFirstPage && (empty($page) || $page < 2)) {
+			$request->getDispatcher()->handle404();
+		}
+
 		$templateMgr = TemplateManager::getManager($request);
 		$this->setupTemplate($request);
 		$press = $request->getPress();
 
 		// Fetch the monographs to display
+		$itemsPerPage = $press->getData('itemsPerPage');
+		$rangeInfo = new DBResultRange($itemsPerPage, $page);
 		$publishedMonographDao = DAORegistry::getDAO('PublishedMonographDAO');
-		$publishedMonographs = $publishedMonographDao->getByPressId($press->getId());
-		$templateMgr->assign('publishedMonographs', $publishedMonographs->toAssociativeArray());
+		$publishedMonographs = $publishedMonographDao->getByPressId($press->getId(), null, $rangeInfo, null, null, null, null, true);
 
-		// Expose the featured monograph IDs and associated params
-		$featureDao = DAORegistry::getDAO('FeatureDAO');
-		$featuredMonographIds = $featureDao->getSequencesByAssoc(ASSOC_TYPE_PRESS, $press->getId());
-		$templateMgr->assign('featuredMonographIds', $featuredMonographIds);
+		// The DAOResultFactory will return the last available page of results. If
+		// it's less than the requested page, we should not display any results.
+		if ($publishedMonographs->page < $page) {
+			$request->getDispatcher()->handle404();
+		}
+
+		if ($page === 1) {
```

Comment:
It seems a little weird to be excluding features from the pagination, but that's just me. I guess we could consider having a single getter that'll fetch both featured and unfeatured monographs, ordering features first, so that both could be paged at the same time? That would simplify some of the logic here too. It would require the templates to check featured status and present them accordingly.

Reply:
You're right. I'm now using a single getter anyway so there's no need to restrict this.